### PR TITLE
APP-3032: Spring WebClient implementation

### PR DESCRIPTION
**Pull request by symphony-hong**
### Ticket
[APP-3032](https://perzoinc.atlassian.net/browse/APP-3032)

### Description
Spring WebClient implementation

### Dependencies
List the other pull requests that should be merged before/along this one.

### Checklist
- [x] Referenced a ticket in the PR title and in the corresponding section
- [x] Filled properly the description and dependencies, if any
- [x] Unit tests updated or added
- [x] Javadoc added or updated
- [ ] Updated the documentation in [docs folder](../docs)

### CODE REVIEW

File: symphony-bdk-http/symphony-bdk-http-webclient/build.gradle [hunk 1]
```diff
@@ -0,0 +1,9 @@
+description = 'Symphony Java BDK Core Http Spring WebClient'
+
+dependencies {
+    implementation project(':symphony-bdk-http:symphony-bdk-http-api')
+
+    implementation 'org.apiguardian:apiguardian-api'
+    implementation 'org.springframework.boot:spring-boot-starter-webflux'
+    implementation 'org.projectreactor:reactor-spring:1.0.1.RELEASE'
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/build.gradle:8]: the version should probably part of the BOM project instead

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  private String temporaryFolderPath;
+
+  public ApiClientBuilderWebClient() {
+    basePath = "";
+    keyStoreBytes = null;
+    keyStorePassword = null;
+    trustStoreBytes = null;
+    trustStorePassword = null;
+    defaultHeaders = new HashMap<>();
+    connectionTimeout = 15000;
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java:47]: for readibility and for large numbers the _ can be used as a separator

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    readTimeout = 60000;
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java:48]: perhaps we should share default timeouts for all our implementations at the API level?

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  private String keyStorePassword;
+  private byte[] trustStoreBytes;
+  private String trustStorePassword;
+  private int connectionTimeout;
+  private int readTimeout;
+  private String temporaryFolderPath;
+
+  public ApiClientBuilderWebClient() {
+    basePath = "";
+    keyStoreBytes = null;
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java:42]: we can avoid setting the fields to null as it is already the case

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+          final KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+          final KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+          keyStore.load(new ByteArrayInputStream(this.keyStoreBytes), this.keyStorePassword.toCharArray());
+          keyManagerFactory.init(keyStore, this.keyStorePassword.toCharArray());
+          builder.keyManager(keyManagerFactory);
+        }
+      }
+      return builder.build();
+    } catch (KeyStoreException | IOException | NoSuchAlgorithmException | CertificateException | UnrecoverableKeyException e) {
+      return null;
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientBuilderWebClient.java:162]: should we throw it as a runtime exception to be able to find out the root cause?

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        return contentType;
+      }
+    }
+    return contentTypes[0];
+  }
+
+  @Override
+  public String escapeString(String str) {
+    try {
+      return URLEncoder.encode(str, "utf8").replaceAll("\\+", "%20");
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:282]: is that all that needs to be escaped? maybe there is something in spring already doing that?

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+      if (!headerParams.containsKey(key)) {
+        String value = defaultHeaderParam.getValue();
+        if (value != null) {
+          requestBodySpec = requestBodySpec.header(key, value);
+        }
+      }
+    }
+
+    if (formParams != null) {
+      if (contentType.startsWith("multipart/form-data")) {
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:103]: there might be some constants in Spring for that

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+        requestBodySpec.body(BodyInserters.fromFormData(formValueMap));
+      }
+    }
+    if (body != null) {
+      requestBodySpec.body(BodyInserters.fromValue(body));
+    }
+
+    Mono<ClientResponse> mono = requestBodySpec.exchange();
+    ClientResponse response = mono.block();
+    Map<String, List<String>> headers = new HashMap<>();
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:128]: we could stream here to collect the headers

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+  public ApiClientWebClient(final WebClient webClient, String basePath, Map<String, String> defaultHeaders) {
+    this.webClient = webClient;
+    this.basePath = basePath;
+    this.defaultHeaderMap = new HashMap<>(defaultHeaders);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> ApiResponse<T> invokeAPI(
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:43]: I would name it invokeApi

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+        b.append(o);
+      }
+      return b.toString();
+    } else {
+      return String.valueOf(param);
+    }
+  }
+
+  @Override
+  public List<Pair> parameterToPairs(String collectionFormat, String name, Object value) {
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:188]: I don't understand this method (and how it is going to be used), could you give a bit of context?

File: symphony-bdk-http/symphony-bdk-http-webclient/build.gradle [hunk 11]
```diff
@@ -0,0 +1,9 @@
+description = 'Symphony Java BDK Core Http Spring WebClient'
```
Comment (symphony-thibault) [inline symphony-bdk-http/symphony-bdk-http-webclient/build.gradle:1]: Can we configure Jacoco for this module?

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java (hunk 10 above)
Comment (symphony-thibault) [reply symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:188]: It is unfortunately used by generated code

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java (hunk 9 above)
Comment (symphony-thibault) [reply symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:43]: Used by generator code, we cannot change the contract. Well, yes we could but we would need to modify the mustache templates

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java (hunk 6 above)
Comment (symphony-hong) [reply symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:282]: I'm not sure about this one. I just took the logic from the implementation of jersey2

File: symphony-bdk-http/symphony-bdk-http-webclient/build.gradle (hunk 11 above)
Comment (symphony-hong) [reply symphony-bdk-http/symphony-bdk-http-webclient/build.gradle:1]: Yes we can but I haven't imagined how we will write unittest for this module

Comment (symphony-youri) [reply symphony-bdk-http/symphony-bdk-http-webclient/build.gradle:1]: Using mockserver? (or wiremock which is nice too)

**Review by symphony-thibault — APPROVED** [review]
LGTM 👍

**Review by symphony-elias — APPROVED** [review]
LGTM

**Pull request merged**